metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanRetentionService.java

```diff
@@ -0,0 +1,193 @@
+
+    if (largestVersion < retention.getMaxVersions()) {
+      return false;
+    }
+    querySoFar.le(EbeanAspectV2.VERSION_COLUMN, largestVersion - retention.getMaxVersions() + 1);
+    return true;
+  }
+
+  private boolean applyTimeBasedRetention(@Nonnull final ExpressionList<EbeanAspectV2> querySoFar,
+      @Nonnull final TimeBasedRetention retention) {
+    querySoFar.lt(EbeanAspectV2.CREATED_ON_COLUMN,
+        new Timestamp(_clock.millis() - retention.getMaxAgeInSeconds() * 1000));
+    return true;
+  }
+
+  @Override
+  @WithSpan
+  public void batchApplyRetention(@Nullable String entityName, @Nullable String aspectName) {
+    log.debug("Applying retention to all records");
+    int numCandidates = queryCandidates(entityName, aspectName).findCount();
+    log.info("Found {} urn, aspect pair with more than 1 version", numCandidates);
+    Map<String, DataHubRetentionInfo> retentionPolicyMap = getRetentionPolicies();
+
+    int start = 0;
+    while (start < numCandidates) {
+      log.info("Applying retention to pairs {} through {}", start, start + _batchSize);
+      PagedList<EbeanAspectV2> rows = getPagedAspects(entityName, aspectName, start, _batchSize);
+
+      try (Transaction transaction = _server.beginTransaction()) {
+        transaction.setBatchMode(true);
+        transaction.setBatchSize(_batchSize);
+        for (EbeanAspectV2 row : rows.getList()) {
+          // Only run for cases where there's multiple versions of the aspect
+          if (row.getVersion() == 0) {
+            continue;
+          }
+          // 1. Extract an Entity type from the entity Urn
+          Urn urn;
+          try {
+            urn = Urn.createFromString(row.getUrn());
+          } catch (Exception e) {
+            log.error("Failed to serialize urn {}", row.getUrn(), e);
+            continue;
+          }
+          final String aspectNameFromRecord = row.getAspect();
+          // Get the retention policies to apply from the local retention policy map
+          List<Retention> retentionPolicies = getRetentionKeys(urn.getEntityType(), aspectNameFromRecord).stream()
+              .map(key -> retentionPolicyMap.get(key.toString()))
+              .filter(Objects::nonNull)
+              .findFirst().<List<Retention>>map(DataHubRetentionInfo::getRetentionPolicies).orElse(
+                  Collections.emptyList());
+          applyRetention(urn, aspectNameFromRecord, retentionPolicies,
+              Optional.of(new RetentionContext(Optional.of(row.getVersion()))));
+        }
+        transaction.commit();
+      }
+
+      start += _batchSize;
+    }
+
+    log.info("Finished applying retention to all records");
+  }
+
+  private Map<String, DataHubRetentionInfo> getRetentionPolicies() {
+    return _server.find(EbeanAspectV2.class)
+        .select(String.format("%s, %s, %s", EbeanAspectV2.URN_COLUMN, EbeanAspectV2.ASPECT_COLUMN,
+            EbeanAspectV2.METADATA_COLUMN))
+        .where()
```

Comment:
hopefully this query will not be super expensive. i was planning to migrate policies away from using a query like this to using elastic directly.. i guess if its once a boot up maybe not too bad

Reply:
better than for policies

Reply:
Right. From tracing, I saw that these queries are sub-10ms queries, since we start with urn, aspect.